experimental/sql/src/alg_to_ssa.py

```diff
@@ -0,0 +1,129 @@
+# Licensed under the Apache License v2.0 with LLVM Exceptions.
+# See https://llvm.org/LICENSE.txt for license information.
+# SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+from dataclasses import dataclass
+from xdsl.ir import Operation, MLContext, Region, Block
+from typing import List, Type, Optional
+from xdsl.dialects.builtin import ArrayAttr, StringAttr, ModuleOp, IntegerAttr
+
+from xdsl.pattern_rewriter import RewritePattern, GreedyRewritePatternApplier, PatternRewriteWalker, PatternRewriter, op_type_rewrite_pattern
+
+import dialects.rel_alg as RelAlg
+import dialects.rel_ssa as RelSSA
+
+# This file contains the rewrite infrastructure to translate the relational
+# algebra dialect to the relational SSA dialect. The current design has a parent
+# class `RelAlgRewriter` that contains functions used for several `Rewriter`s.
+# All other `Rewriter`s inherit from that class. To introduce SSA, the passes
+# insert the children, that should become operands, before the matched op, then
+# use the `added_operations_before` functionality of the rewriter to link the
+# operands to the new operation.
+
+
+@dataclass
+class RelAlgRewriter(RewritePattern):
+
+  def convert_datatype(self, type_: RelAlg.DataType) -> RelSSA.DataType:
+    if isinstance(type_, RelAlg.String):
+      return RelSSA.String.get(type_.nullable)
+    if isinstance(type_, RelAlg.Int32):
+      return RelSSA.Int32()
+    raise Exception(
+        f"datatype conversion not yet implemented for {type(type_)}")
+
+  def lookup_type_in_schema(self, name: str,
+                            bag: RelSSA.Bag) -> Optional[RelSSA.DataType]:
+    for s in bag.schema.data:
+      if s.elt_name.data == name:
+        return s.elt_type
+    return None
+
+
+#===------------------------------------------------------------------------===#
+# Expressions
+#===------------------------------------------------------------------------===#
+
+
+@dataclass
+class LiteralRewriter(RelAlgRewriter):
+
+  @op_type_rewrite_pattern
+  def match_and_rewrite(self, op: RelAlg.Literal, rewriter: PatternRewriter):
+    new_op = RelSSA.Literal.get(op.val, self.convert_datatype(op.type))
+    rewriter.insert_op_before_matched_op(
+        RelSSA.Literal.get(op.val, self.convert_datatype(op.type)))
+    rewriter.erase_matched_op()
+
+
+@dataclass
+class ColumnRewriter(RelAlgRewriter):
+
+  @op_type_rewrite_pattern
+  def match_and_rewrite(self, op: RelAlg.Column, rewriter: PatternRewriter):
+    res_type = self.lookup_type_in_schema(
+        op.col_name.data,
+        op.parent_op().parent_op().input.op.results[0].typ)
```

Comment:
I am not happy with this particular line just yet. This depends on the fact that the parent_op of the parent_op is a select and it cannot be used otherwise. I am still considering options of how to circumvent this. 

I am thinking of adding a data structure that keeps track of the schemas of different tables. Then, this would just become a lookup in that data structure. But how the data structure looks like is not clear either. I feel that a dictionary mapping table_names (strings) to dictionaries mapping column_names (strings) to data types could be feasible, but then we need the table_name here, i.e. we basically have the exact same problem.

In the end, I feel that I cannot implement this without having to lookup the parent_op and get the schema from there, in a sense, that is the price we pay for not having the table_name in the column operation.

Obviously, the other option we have is going back through the entire pipeline and adding the table_name everywhere. Then the dictionary approach would work again.

I plan to cleanup this parent_lookup version in a cleaner way s.t. it generalizes better (i.e. lookup the next parent_op with an input table and lookup in its schema), but I would happy to hear suggestions.

Reply:
Ha, I didn't see this comment during my review because you had clean-up this line already. I think the clean-up was a good step in the right direction, but see my comments there.